### PR TITLE
Fix Collective <-> Elections initialization.

### PR DESCRIPTION
**Pull request by kianenigma**
Fix https://github.com/paritytech/substrate/issues/5097. 

Although, this mostly demonstrates the issue. This is kind of a footgun since the genesis value of collective and elections must be the same, otherwise the behaviour is even more obscure.  and I think a better way is to: 

- `trait ChangeMembers` should expose one function `initial_members()` that `elections()` can call and get the initial members. 

Basically, from the PoV of elections, whatever module we report the new members to, also gives us the initial ones. Hence, no need for a new trait and can be within the same trait.  

Although, this brings the question of: How much stake should we give to the initial members? the collective module does not know that. The current genesis approach is good in the sense that it makes it configurable. 

polkadot companion: https://github.com/paritytech/polkadot/pull/954

--- 

As noted also by @jacogr, polkadot names the elections module `ElectionPhragmen` while substrate master does `Elections`. Why is this? if no particular reason, would be better to have this breaking (for the UI) change now before 2.0? Fixing such stuff should make it easier for the UI to support both FF and Polkadot/Kusama.

### CODE REVIEW

**Comment by kianenigma** [comment]
> Yeah, as you already said, you should inform collective about the genesis members.

Do you have an idea on how much tokens to lock then? we _could_ make even that configurable by a `type InitialMembersStakeRate: Get<PerThing>`.

**Comment by jacogr** [comment]
Renaming won't be breaking for the API/UI, it has to support both, so renaming is fine. (It will continue supporting both as wel for at least the medium terml, since there are networks with both). 

Having said that, I'm in favor of cleaning up the naming there, at least for sanity - which also means that "at some point in the future" the UIs can start sticking to one. (And it certainly affects other external teams as well supporting collective, e.g. Commonwealth also has checks for both naming types)

**Comment by kianenigma** [comment]
Or, I think a more logical way is to disallow collective to have genesis config, move the genesis to elections and then pass it to the collective in `add_extra_gensis` of elections.

**Comment by bkchr** [comment]
> Renaming won't be breaking for the API/UI, it has to support both, so renaming is fine. (It will continue supporting both as wel for at least the medium terml, since there are networks with both).
> 
> Having said that, I'm in favor of cleaning up the naming there, at least for sanity - which also means that "at some point in the future" the UIs can start sticking to one. (And it certainly affects other external teams as well supporting collective, e.g. Commonwealth also has checks for both naming types)

This just sounds like you wan to have some sort of unique identifier per pallet, which I would not make dependent on the name the user chooses in `construct_runtime!`

**Comment by bkchr** [comment]
> > Yeah, as you already said, you should inform collective about the genesis members.
> 
> Do you have an idea on how much tokens to lock then? we _could_ make even that configurable by a `type InitialMembersStakeRate: Get<PerThing>`.
> 
> I'll go with entire balance for now.

Why not pass this into genesis as well? Add an `extra_config` that contains the member ids + locked.

**Comment by kianenigma** [comment]
> Why not pass this into genesis as well? Add an extra_config that contains the member ids + locked.

That would be bad because we'd have to pass the stake to the genesis of membership, and membership has no notion of stake or balance. It just knows members accountid. 

 To avoid this and the need to add an extra function to the trait, the genesis members, with their stake, should be set in `elections`, and then it is passed to whatever is `T::MembersChanged`. Will also add a warning to the genesis of `collective` that if you are using it with `elections`, you should NOT set them directly in the genesis value of `collective`. To make it real safe, we could actually make it not-configurable, but that would be bad because some people might want to have a static set of members and actually not use elections module here.

**Comment by kianenigma** [comment]
Updated and I am pretty happy with the approach now. The initial members can be set either by setting genesis of elections or directly in collective's genesis.

**Comment by bkchr** [comment]
Missing Polkadot pr.

File: frame/elections-phragmen/src/lib.rs [hunk 1]
```diff
@@ -160,11 +163,44 @@ decl_storage! {
...
+				);
+
+				// reserve candidacy bond and set as members.
+				T::Currency::reserve(&member, T::CandidacyBond::get())
+					.expect("Genesis member does not have enough balance to be a candidate");
+
+				// Note: all members will only vote for themselves, hence they must be given exactly
+				// their own stake as total backing. Any sane election should behave as such.
+				// Nonetheless, stakes will be updated for term 1 onwards according to the election.
+				// TODO: can't use append here?
```
Comment (bkchr) [inline frame/elections-phragmen/src/lib.rs]: Todo

File: frame/elections-phragmen/src/lib.rs [hunk 2]
```diff
@@ -160,11 +163,44 @@ decl_storage! {
...
 		/// can never enter this vector and is always implicitly assumed to be a candidate.
 		pub Candidates get(fn candidates): Vec<T::AccountId>;
+	} add_extra_genesis {
+		config(members): Vec<(T::AccountId, BalanceOf<T>)>;
+		build(|config: &GenesisConfig<T>| {
+			let members = config.members.iter().map(|(ref member, ref stake)| {
+				// make sure they have enough stake
+				assert!(
+					T::Currency::free_balance(member) >= *stake,
+					"Genesis member does not have enough stake"
```
Comment (bkchr) [inline frame/elections-phragmen/src/lib.rs]:
```suggestion
					"Genesis member does not have enough stake",
```

**Comment by kianenigma** [comment]
Companion added.

**Comment by bkchr** [comment]
Polkadot job runs after all other are successful. Please fix merge conflicts.

**Comment by bkchr** [comment]
Still needs some love.

**Comment by kianenigma** [comment]
made some love just now.

**Comment by kianenigma** [comment]
transaction pool :( https://gitlab.parity.io/parity/substrate/-/jobs/441783

**Pull request merged**